packages/neuron-wallet/src/services/indexer/queue.ts

```diff
@@ -0,0 +1,221 @@
+
+enum TxPointType {
+  CreatedBy = 'createdBy',
+  ConsumedBy = 'consumedBy',
+}
+
+export default class Queue {
+  private lockHashes: string[]
+  private indexerRPC: IndexerRPC
+  private getBlocksService: GetBlocks
+  private per = 50
+  private interval = 1000
+  private blockNumberService: BlockNumber
+  private tipNumberListener: Subscription
+  private tipBlockNumber: bigint = BigInt(-1)
+
+  private stopped = false
+  private indexed = false
+
+  private inProcess = false
+
+  private resetFlag = false
+
+  constructor(url: string, lockHashes: string[], tipNumberSubject: Subject<string | undefined>) {
+    this.lockHashes = lockHashes
+    this.indexerRPC = new IndexerRPC(url)
+    this.getBlocksService = new GetBlocks()
+    this.blockNumberService = new BlockNumber()
+    this.tipNumberListener = tipNumberSubject.subscribe(async (num: string) => {
+      if (num) {
+        this.tipBlockNumber = BigInt(num)
+      }
+    })
+  }
+
+  public setLockHashes = (lockHashes: string[]): void => {
+    this.lockHashes = lockHashes
+    this.indexed = false
+  }
+
+  public reset = () => {
+    this.resetFlag = true
+  }
+
+  /* eslint no-await-in-loop: "off" */
+  /* eslint no-restricted-syntax: "off" */
+  public start = async () => {
+    while (!this.stopped) {
+      try {
+        this.inProcess = true
+        if (this.resetFlag) {
+          await this.blockNumberService.updateCurrent(BigInt(0))
+          this.resetFlag = false
+        }
+        const { lockHashes } = this
+        const currentBlockNumber: bigint = await this.blockNumberService.getCurrent()
+        if (!this.indexed || currentBlockNumber !== this.tipBlockNumber) {
+          if (!this.indexed) {
+            await this.indexLockHashes(lockHashes)
+            this.indexed = true
+          }
+          const minBlockNumber = await this.getCurrentBlockNumber(lockHashes)
+          for (const lockHash of lockHashes) {
+            await this.pipeline(lockHash, TxPointType.CreatedBy, currentBlockNumber)
+          }
+          for (const lockHash of lockHashes) {
+            await this.pipeline(lockHash, TxPointType.ConsumedBy, currentBlockNumber)
+          }
+          if (minBlockNumber) {
+            await this.blockNumberService.updateCurrent(minBlockNumber)
+          }
+        }
+        await this.yield(this.interval)
+      } catch (err) {
+        if (err.message.startsWith('connect ECONNREFUSED')) {
+          logger.debug('sync indexer error:', err)
+        } else {
+          logger.error('sync indexer error:', err)
+        }
+      } finally {
+        await this.yield()
+        this.inProcess = false
+      }
+    }
+  }
+
+  public processFork = async () => {
+    while (!this.stopped) {
+      try {
+        const tip = this.tipBlockNumber
+        const txs = await IndexerTransaction.txHashes()
+        for (const tx of txs) {
+          const result = await this.getBlocksService.getTransaction(tx.hash)
+          if (!result) {
```

Comment:
Under what conditions will tx not exist?

Reply:
The block of tx not in the longest chain now.